### PR TITLE
<Timeline/> - new component

### PR DESCRIPTION
**Pull request by adimorwix**

### CODE REVIEW

File: src/Timeline/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import * as React from 'react';
+
+export interface TimelineItem {
+  id: string | number;
+  label: string;
+  labelAction?: React.ReactNode;
+  customPrefix?: React.ReactNode;
+  suffix?: React.ReactNode | string;
```
Comment (EgoziE) [inline src/Timeline/index.d.ts:8]: string is covered inside React.ReactNode, can be removed

File: src/Timeline/Timeline.uni.driver.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { baseUniDriverFactory } from 'wix-ui-test-utils/base-driver';
+import { dataHooks } from './constants';
+
+export const timelineDriverFactory = (base, body) => {
+  return {
+    ...baseUniDriverFactory(base, body),
+
+    /** Is the timeline exists */
+    isTimelineExists: () =>
```
Comment (shlomitc) [inline src/Timeline/Timeline.uni.driver.js:9]: `baseUniDriverFactory` already has `exists()` so I think you can remove it

File: src/Timeline/Timeline.uni.driver.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      base.$(`[data-hook="${dataHooks.timelineList}"]`).exists(),
+
+    /** Get the label text by item id */
+    getLabelText: async itemsId =>
```
Comment (shlomitc) [inline src/Timeline/Timeline.uni.driver.js:13]: rename to `getLabelTextAt(idx)`
where `idx` is the position of it, instead of an id

File: src/Timeline/Timeline.uni.driver.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      await base
+        .$(`[data-hook="${dataHooks.timelineLabel}-${itemsId}"]`)
+        .text(),
+
+    /** Get the suffix text by item id */
+    getSuffixText: async itemsId =>
```
Comment (shlomitc) [inline src/Timeline/Timeline.uni.driver.js:19]: same comment as `getLabelText`

File: src/Timeline/Timeline.uni.driver.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      base.$(`[data-hook="${dataHooks.timelineLabelAction}-${itemsId}"]`),
+
+    /** Is default prefix wrapper by item id */
+    isDefaultPrefixExists: async itemsId =>
+      await base
+        .$(`[data-hook="${dataHooks.timelineDefaultPrefix}-${itemsId}"]`)
+        .exists(),
+
+    /** label action wrapper by item id */
+    getCustomPrefixElement: itemsId =>
```
Comment (shlomitc) [inline src/Timeline/Timeline.uni.driver.js:39]: same comment as `getLabelActionElement`

File: src/Timeline/Timeline.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  /** Applied as data-hook HTML attribute that can be used in the tests */
+  dataHook: PropTypes.string,
+
+  /** A css class to be applied to the component's root element */
+  className: PropTypes.string,
+
+  /** timeline events items */
+  items: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
```
Comment (shlomitc) [inline src/Timeline/Timeline.js:106]: I think an `id` is not needed, we can simply use the index of the item

File: src/Timeline/Timeline.js [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import styles from './Timeline.st.css';
+import { dataHooks } from './constants';
+
+/** A timeline is a display of a list of events */
+class Timeline extends React.PureComponent {
+  state = {};
+
+  _isString = a => typeof a === 'string';
+
+  _getListItem(item) {
```
Comment (shlomitc) [inline src/Timeline/Timeline.js:14]: can you please split this to a `<TimeLineItem/>` component?
When a method on a component becomes that big, I prefer it separated

File: src/Timeline/Timeline.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            <div
+              data-hook={`${dataHooks.timelineCustomPrefix}-${item.id}`}
+              className={styles.prefixCustom}
+            >
+              {item.customPrefix}
+            </div>
+          ) : (
+            <div
+              data-hook={`${dataHooks.timelineDefaultPrefix}-${item.id}`}
+              className={styles.prefixDefault}
```
Comment (shlomitc) [inline src/Timeline/Timeline.js:32]: let's call it `bulletIndicator` instead of prefixDefault

File: src/Timeline/Timeline.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        </div>
+        <div className={styles.label}>
+          <Text
+            dataHook={`${dataHooks.timelineLabel}-${item.id}`}
+            weight="normal"
+            size="small"
+          >
+            {item.label}
+          </Text>
+          <span> </span>
```
Comment (shlomitc) [inline src/Timeline/Timeline.js:44]: I think we can remove this `span` and just use some margin on the `labelAction` class

File: src/Timeline/docs/index.story.js [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+              '    labelAction: (\n' +
+              '      <TextButton weight="normal" size="small">\n' +
+              '        View Order\n' +
+              '      </TextButton>\n' +
+              '    ),\n' +
+              '  },\n' +
+              ']}/>',
+          }),
+
+          code({
```
Comment (shlomitc) [inline src/Timeline/docs/index.story.js:142]: you can delete this one

File: src/Timeline/Timeline.js (hunk 9 above)
Comment (adimorwix) [reply src/Timeline/Timeline.js:44]: If only the action down a line we have unnecessary space with a margin

File: .wuf/components.json [hunk 11]
```diff
@@ -424,5 +424,8 @@
   },
   "VerticalTabsItem": {
     "path": "src/VerticalTabsItem"
+  },
+  "Timeline": {
```
[the remark this answers is not in the document]
Comment (EgoziE) [reply .wuf/components.json:428]: @shlomitc https://github.com/wix/wix-style-react/pull/5390

File: src/Timeline/Timeline.js (hunk 9 above)
Comment (shlomitc) [reply src/Timeline/Timeline.js:44]: so we can add `margin-right` to the text, so it will break one word earlier

File: src/Timeline/Timeline.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+  /** Applied as data-hook HTML attribute that can be used in the tests */
+  dataHook: PropTypes.string,
+
+  /** A css class to be applied to the component's root element */
+  className: PropTypes.string,
+
+  /** timeline events items */
+  items: PropTypes.arrayOf(
+    PropTypes.shape({
+      /** event text */
```
Comment (shlomitc) [inline src/Timeline/Timeline.js:106]: timeline item textual description

File: src/Timeline/index.d.ts (hunk 1 above)
Comment (adimorwix) [reply src/Timeline/index.d.ts:8]: In labelAction too for example?
I think type make it more clear

File: src/Timeline/Timeline.st.css [hunk 13]
```diff
@@ -0,0 +1,82 @@
+:import {
+  -st-from: "../Foundation/stylable/colors.st.css";
+  -st-named: D30,D80;
+}
+
+.root {
+  text-align: center;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:7]: not needed

File: src/Timeline/Timeline.st.css [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+  background-color: value(D80);
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:8]: not needed.
Usually, this component and similar widgets will appear on top of a Card component. We can add another example in the playground that uses `<Card/>` and `<Card.Content/>`

File: src/Timeline/Timeline.st.css [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  padding-left: 23px;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:9]: not needed

File: src/Timeline/Timeline.st.css [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  padding-right: 30px;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:10]: not needed

File: src/Timeline/Timeline.st.css [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+  padding-top: 28px;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:11]: not needed

File: src/Timeline/Timeline.st.css [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+.timeline {
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:14]: all of these are not needed besides the `list-style: none;`
Maybe we can just merge this with `root` and remove the `div` from the markup?

File: src/Timeline/Timeline.st.css [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+  list-style: none;
+  text-align: left;
+  padding: 0;
+  margin: 0;
+}
+
+.timeline .event {
+  display: list-item;
+  position: relative;
+  padding-bottom: 30px;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:26]: This should appear for all besides the last child

File: src/Timeline/Timeline.st.css [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+  border-left: 1px solid value(D30);
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:27]: it's an interesting solution, but this means that the prefix should cover it. I guess it will cause some small issues in CSS so let's see if we can overcome them in an elegant way

File: src/Timeline/Timeline.st.css [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+  display: block;
+  position: relative;
+  list-style: none;
+  text-align: left;
+  padding: 0;
+  margin: 0;
+}
+
+.timeline .event {
+  display: list-item;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:24]: what's the point of using list-item if we don't actually display it? maybe it should be semantically used as `ul` and `li` in the markup, this will lead to better a11y

File: src/Timeline/Timeline.st.css [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+  padding-bottom: 30px;
+  border-left: 1px solid value(D30);
+}
+
+.timeline .event:last-child {
+  border-left: none;
+}
+
+.timeline .event .prefix{
+  width: 3%;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:35]: I don't think we use percentages in the values of items. it's either fixed pixel size or "flex" containers, please find the right solution following this description
image

File: src/Timeline/Timeline.st.css [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+  display: inline-block;
+}
+
+.timeline .event .prefix .prefixDefault{
+  position: absolute;
+  display: block;
+  background: value(D30);
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:42]: background should be transparent

File: src/Timeline/Timeline.st.css [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+.timeline .event:last-child {
+  border-left: none;
+}
+
+.timeline .event .prefix{
+  width: 3%;
+  display: inline-block;
+}
+
+.timeline .event .prefix .prefixDefault{
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:39]: let's split this, there should be a standalone `bulletIndicator` which is the component styles (width, height, border-radius, ...) and there should be 

`.prefix` that takes care of the positioning relative to the parent node. both classes should be applied to the HTML node.

File: src/Timeline/Timeline.st.css [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+  position: absolute;
+  display: block;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:41]: isn't it block by default?

File: src/Timeline/Timeline.st.css [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+  left: -4px;
+}
+
+.timeline .event .prefix .prefixCustom:first-child{
+  position: absolute;
+  display: block;
+  border-radius: 50%;
+  content: "";
+  top:-1px;
+  box-shadow: 0 0 0 1px value(D80);
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:58]: looks strange, why is this needed?

File: src/Timeline/Timeline.st.css [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+  content: "";
+  top: 4.5px;
+  box-shadow: 0 0 0 5px value(D80);
+  left: -4px;
+}
+
+.timeline .event .prefix .prefixCustom:first-child{
+  position: absolute;
+  display: block;
+  border-radius: 50%;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:55]: I don't think this should be, this should be the responsibility of the custom prefix node which is rendered by the consumer

File: src/Timeline/Timeline.st.css [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+  content: "";
+  top:-1px;
+  box-shadow: 0 0 0 1px value(D80);
+  background-color: value(D80);
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:59]: not needed

File: src/Timeline/Timeline.st.css [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+  left: -12px;
+}
+
+.timeline .event .label{
+  width: 67%;
```
Comment (shlomitc) [inline src/Timeline/Timeline.st.css:64]: fix according to spec constraints

File: src/Timeline/Timeline.st.css (hunk 19 above)
Comment (adimorwix) [reply src/Timeline/Timeline.st.css:17]: only the padding-bottom and border-left should not exist in  last-child

File: src/Timeline/Timeline.st.css (hunk 23 above)
Comment (adimorwix) [reply src/Timeline/Timeline.st.css:34]: we can't use transparent color. the color cover the border

File: src/Timeline/Timeline.st.css (hunk 28 above)
Comment (adimorwix) [reply src/Timeline/Timeline.st.css:50]: the color cover the border

File: src/Timeline/Timeline.st.css (hunk 18 above)
Comment (adimorwix) [reply src/Timeline/Timeline.st.css:7]: We have to override the defaults

File: src/Timeline/TimelineItem.st.css [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+.event .suffix{
+  width: 25%;
+  display: inline-block;
+  text-align: right;
+  vertical-align: top;
+  padding-top: 3px;
+}
+
+.event .labelAction{
+  display: inline-block;
```
[the remark this answers is not in the document]
Comment (adimorwix) [reply src/Timeline/TimelineItem.st.css:63]: the action should be in the same line

**Comment by ayeletr** [comment]
Some fixes needed:
Suffix - align to content top, and adjust color.
Remove for now the different prefix, we will define them when we’ll have specific requests.

**Comment by milkyfruit** [comment]
Why `label` is string and `suffix` is a node?
Can't it all be made as a node?

Also, I think the first example should be more simple showing the main structure. e.g.:
```
<Timeline items={[
  {label:'Item 1',},
  {label:'Item 2',}
  ]}
/>
```

**Comment by shlomitc** [comment]
@milkyfruit - The current requirements were for a textual label, it is also important to keep a line-height of 18px string in order to align the "bullet" and make the line stretch.
If we decide on having an open node, this will change the component structure, so I prefer not doing so until we have an actual need.

In any case, let's open a JIRA task related to documentation fixes, I think the code is ready to merge and I don't want to block it for this
